### PR TITLE
Player

### LINKED ISSUE #8: Rotate Tile Functions
- [ ] clockwise
- [ ] counter clockwise

### PR DESCRIPTION
**Pull request by tornglintaffychen**
completed player functions (including tiles, points, and spaces functions) 
closes #8, #5, #4 , 
ready for code review

### CODE REVIEW

File: tiles.js [hunk 1]
```diff
@@ -57,7 +57,7 @@ var tiles = [{
     paths: [4, 3, 6, 1, 0, 7, 2, 5]
 }, {
     image: "",
-    paths: [2, 6, 0, 4, 3, 7, 1, 5]
```
Comment (ashes74) [inline tiles.js]: This is an invalid tile.

[3, 7, 4, 0, 6, 5, 1]

should it be: [3, 4, 7, 0, 1, 6, 5, 2] ?

File: player.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    return Player.placeTile(minusTwo);
+}
+
+// when we get the tile, it is already rotated or not, from angular.
+Player.prototype.placeTile = function (tile) {
+    var index = this.tiles.indexOf(tile)
+    this.tiles.splice(index, 1);
+
+    for (var i = 0; i < tile.length; i++) {
+        this.nextSpace.points[i].neighbors.push(points[tile[i]]);
```
Comment (ashes74) [inline player.js:56]: Where is points defined for (points[tile[i]]) to work?

File: player.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    let playersWhoCanPlay = Game.players.filter(function (player) {
```
Comment (ashes74) [inline player.js]: Player should not be have to generate the  playersWhoCanPlay array. This should be handled by the Game object

File: player.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        return player.canPlay;
+    });
+
+    // after a player placed a tile, we check all canPlay players to use keepMoving function to see if can move and/or die
+    playersWhoCanPlay.forEach(function (player) {
```
Comment (ashes74) [inline player.js]: Let's move this logic to to the main controller that is handling events in the game (in Angular I think )

File: player.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    //function need to be written in the game side for when players are dead b/c of the placed tile
+    //OR we can just keep the dead players on the point where they end up dying at b/c no other players would ever hit that point.
+    Game.removeAllDeadPlayersForThisPlaceTile();
+}
+
+
+// if a plyer is on the edge, then no need to run this function
+function newSpace(oldSpace) {
+    if (this.nextSpacePointsIndex === 0 && this.nextSpacePointsIndex === 1) {
```
Comment (ashes74) [inline player.js]: These conditions will never be met. Should they be '||'?

File: player.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    } else if (this.nextSpacePointsIndex === 4 && this.nextSpacePointsIndex === 5) {
+        return board[oldSpace.y + 1][oldSpace.x];
+    } else {
+        return board[oldSpace.y][oldSpace.x - 1];
+    }
+};
+
+function moveTo(pointer) {
+    let pointer = pointer;
+    let moveTo = pointer.neighbors.filter(function (neighbor) {
```
Comment (ashes74) [inline player.js]: it's confusing calling the function and variable the same name.

File: player.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        let oldSpace = this.nextSpace;
+        let newSpace = newSpace(oldSpace);
+        this.nextSpace = newSpace;
+
+        checkDeath(this);
+        moveTo(this.point);
+    };
+};
+
+function checkDeath(self) {
```
Comment (ashes74) [inline player.js:110]: Player should not be calling Game functions. 

It should check for collision with a player or the edge and set it's canPlay variable. Game will handle the rest

Collision = hitting a point that has been travelled or that is an edge. 

if(point.travelled || point.edge) Player.canPlay = false;

File: player.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    ) : (
+        Game.players.forEach(function (player) {
+            if (JSON.stringify(player.point) === JSON.stringify(self.point)) {
+                Game.die(self);
+            };
+        });
+    )
+}
+
+Game.prototype.die = function (player) {
```
Comment (ashes74) [inline player.js]: This should be a Player.prototype function

File: player.js [hunk 9]
```diff
@@ -0,0 +1,124 @@
+// remove startingPoint and nextSpace
+function Player(name) {
+    this.name = name;
+    this.marker = null;
+
+    this.point = null; // should be a Point object
+    this.nextSpace = null; // [x, y] // depends on the angular Space.x, Space.y
+
+    // in each Space.points array, find this specific point and get the position (integer) inside this space.
+    this.nextSpacePointsIndex = null;
```
Comment (ashes74) [inline player.js:10]: I don't understand what this is doing

File: player.js (hunk 5 above)
Comment (tornglintaffychen) [inline player.js]: yes, it should be ||, fixed

File: player.js (hunk 6 above)
Comment (tornglintaffychen) [inline player.js]: changed the variable name to nextPoint instead of moveTo

File: player.js (hunk 8 above)
Comment (tornglintaffychen) [inline player.js]: changed.

File: tiles.js (hunk 1 above)
Comment (maxww) [inline tiles.js]: [3, 7, 4, 0, 2, 6, 5, 1], fixed

File: player.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+Player.prototype.keepMoving = function() {
+    let movable = moveTo(this.point);
+    while (movable) {
+        this.point.travelled = true;
+        this.point = moveTo(this.point);
+        let oldSpace = this.nextSpace;
+        let newSpace = newSpace(oldSpace);
+        this.nextSpace = newSpace;
+
+        checkDeath(this);
```
Comment (ashes74) [inline player.js:105]: why isn't this just this.checkDeath()

File: player.js [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    };
+};
+
+function checkDeath(self) {
+    self.point.edge ? (
+        Player.die(self)
+    ) : (
+        Player.players.forEach(function(player) {
+            if (JSON.stringify(player.point) === JSON.stringify(self.point)) {
+                Game.die(self);
```
Comment (ashes74) [inline player.js:116]: A player should only control self. 
If it's going to die. kill itself. 

Game.moveAllPlayers will handle the death of all the others

**Pull request closed without merging**